### ISSUE TITLE
Missing extension when trying to download webpages

### ISSUE THREAD

**Issue by Chaphasilor** [hgouveia/node-downloader-helper#27]
I just tried 'downloading' `https://google.com` using the following command:

```javascript
new DownloaderHelper('https://google.com', '/home/downloads', {
      fileName: {name: 'test', ext: false},
      ...
    });
```

This will download the page, but the filename will be `test.`, without the `html` part. On windows, this will cause a corrupt file which can't be renamed or deleted.

I'm guessing this is because of `ext: false` and the fact that there normally is no extension when loading a webpage.  
However, this should not happen, at least not with the trailing dot.

**Comment by Chaphasilor** [comment]
FYI I also encountered other issues with `ext: false`, for example about 50 random characters as the extension when downloading a video. I don't have the link rn, but maybe you should check your code that determines the extension and provide some sensible fallbacks for common file formats?

**Comment by hgouveia** [comment]
@Chaphasilor thanks for your report!, this is great information I will check it out as soon as possible, also if you already have a solution you are welcome to contribute with a PR

**Comment by Chaphasilor** [comment]
Sadly I don't have too much time at hand right now, but I might take a look at it at some point.  Don't quote me on that though...

**Comment by Chaphasilor** [comment]
Hey @hgouveia it's me again :D
Any progress on the issue?

**Comment by hgouveia** [comment]
> Hey @hgouveia it's me again :D
> Any progress on the issue?

hey @Chaphasilor, to be honest, I didn't check it yet, I will try during next week, sorry about that, I will let you know

**Comment by Chaphasilor** [comment]
@hgouveia hey no worries, just wanted to check, I had almost 0 problems with this in the last couple months, but it probably should be fixed at some point. If you point me in the right direction I might be able to help you with the fix as well, should have some more time in one or two weeks as well! :D

**Comment by hgouveia** [comment]
@Chaphasilor so basically, i have a function that gets the filename based on this HTTP header `content-disposition`, but if not available, it will get it from the URL, so because i am using `path.basename` probably is ignoring the extension, I would need to check this,  https://github.com/hgouveia/node-downloader-helper/blob/master/src/index.js#L595

in theory, is an easy fix

**Comment by Nashorn** [comment]
@hgouveia - how positive are you about getting the file name from **'content-disposition'**? Will this work 100% of the time? If so why not rely on it completely? I've seen URLs that have no ext, or the content/mime types are not correct, for example, trying to download a font from a URL, no ext, mimeType is **'application/octet-stream'**

**Comment by Chaphasilor** [comment]
@Nashorn the `content-disposition` header is not always present, especially if a link isn't normally meant for downloading (like a webpage). There is also the possibility to set the header just to `attachment` but without providing a filename, which I've also seen used before.

We should probably try to use the mime type to determine a valid extension, and if that doesn't work simply skip the extension altogether (making sure there won't be a trailing dot)...

**Comment by hgouveia** [comment]
@Nashorn as @Chaphasilor stated, the issues is that **content-disposition** is not always available so we can't rely only on that,

Trying to determine the mime-type from **content-type** to give a valid extension is a good suggestion but this will make this project to use a third-party library for ex: [mime-types](https://www.npmjs.com/package/mime-types) that I want to avoid, or have a list with known extensions but it won't cover all of them, but since you can provide a callback to change the filename, it could be done from there if you wish so, but I need to update the arguments of this callback to also provided mime-type  something like `{ filename: (fileName, filePath, mimeType) => 'my.file' )};`, 

I still didn't have the time to fully review the issue, sorry, as soon I can I will check it more in-depth

**Comment by Chaphasilor** [comment]
@hgouveia And what if you export an additional method that tries to determine the extension? If that fails, we could (in our respective projects) require the user to also input an extension, or warn them that the resulting file won't have one.  
You don't have to handle every case :)

**Comment by hgouveia** [comment]
> @hgouveia And what if you export an additional method that tries to determine the extension? If that fails, we could (in our respective projects) require the user to also input an extension, or warn them that the resulting file won't have one.
> You don't have to handle every case :)

well in theory that is the aim of the callback that can be provided to the `filename` option, if you provided that, it won't try to determine the filename, I will use the string returned by this callback, this is string should include also the ext, `myfile.html`,  but could be a nice addition to at least emitting an event when it fail , I can imagine something like

```
dl.on('retrive.filename.failed', () => dl.setFilename('myfile.html') );
```

**Comment by Chaphasilor** [comment]
@hgouveia I'm actually thinking of something more flexible. Something you can call *before* the download starts, so that it can be handled properly. If this is done via a callback once the download starts, it severly limits how it can be handled...

What do you think?

**Comment by Chaphasilor** [comment]
@hgouveia I took another look at this. It is possible to use the `getTotalSize()` function in order to get the automatically determined filename before the download is started, so your proposed change (exposing the mime-type) seems reasonable.

I'll supply you with a PR in the coming days that fixes the bug with files ending in a dot, and exposes the mime-type to the callback function.

**Comment by Wetbikeboy2500** [comment]
> > I am currently reading the extension data of the file after downloading to rename it with the correct extension.
> 
> Hey @Wetbikeboy2500 would you mind explaining how you do this over in #27? :)
> 
> _Originally posted by @Chaphasilor in https://github.com/hgouveia/node-downloader-helper/issues/10#issuecomment-774210762_

It has been a bit, but I can explain it. I really did not do anything groundbreaking. The first step is to isolate the file name. I then checked to see if there was a dot within its name and them did a substring from the end of the filename to the dot. 

This issue has a few more things to consider though. When you request a website like google.com by `https://google.com` the browser will actually give you `https://google.com/index.html`. You can see this by doing an inspect source on the page. This heavily depends on how the server is set up and most pages seem to just return html without giving a name or extension, but this is okay since it defaults its type to html since it is a url. Other links embedded in a page also have corresponding tags and attributes so the browser knows what type of content it is and should try to assume it as. 

What I am getting at is that context matters, and you should know the general data type it will be before you even have a url. I was working with images from img tags so I could assume .jpg or .png from the website's standard. 

The `ext: false` issues come from https://github.com/hgouveia/node-downloader-helper/blob/1f2ac3b9441a79c097e55ee73274976f6e251629/src/index.js#L692-#L693

When ext is false, it will assume there is a dot for the extension and then join the name with whatever extension could exist. This could cause there to be a trailing dot with no extension. It also has the other case of popping the actual filename from the end of the list and putting that as the extension.

When it comes to external resources like the html on a website, I think it is more up to the developer to assume extension and data types than the downloading tool they are using. The assets on those pages like SVGs, Images, etc. can be assumed by the developer and can then use mime-types to narrow down data types of those specific assets after download.

**Comment by Chaphasilor** [comment]
Okay, thanks a lot for your insights. I was hoping there was more to it so that this somehow could be made more reliable, but it seems like there just aren't that many possibilities of determining the file extension ://

**Comment by hgouveia** [comment]
Closing, fixed in new version `1.0.18`, thanks to  @Chaphasilor 
https://github.com/hgouveia/node-downloader-helper/releases/tag/v1.0.18

**Comment by Chaphasilor** [comment]
Great! 🦾
